iceoryx_utils/include/iceoryx_utils/cxx/expected.hpp

```diff
@@ -707,33 +701,15 @@ class expected<ValueType, ErrorType>
     [[deprecated]] expected& on_success(const cxx::function_ref<void(ValueType&)>& callable) noexcept;
     expected& and_then(const cxx::function_ref<void(ValueType&)>& callable) noexcept;
 
-    ///
-    /// @brief if the expected contains a success value and its type is a non-empty optional, retrieve the value from
-    ///         the optional and provide it as the argument to the provided callable
-    /// @param[in] callable the callable to be called with the contents of the optional
-    /// @return reference to the expected
-    /// @code
-    ///     anExpectedOptional.and_then([](int& value){
-    ///         std::cout << "the optional contains the value: " << result << std::endl;
-    ///     })
-    /// @endcode
-    ///
-    template <typename Optional = ValueType, typename std::enable_if<is_optional<Optional>::value, int>::type = 0>
-    const expected& and_then(const cxx::function_ref<void(typename Optional::type&)>& callable) const noexcept;
 
-    ///
-    /// @brief if the expected contains a success value and its type is a non-empty optional, retrieve the value from
-    ///         the optional and provide it as the argument to the provided callable
-    /// @param[in] callable the callable to be called with the contents of the optional
-    /// @return reference to the expected
-    /// @code
-    ///     anExpectedOptional.and_then([](int& value){
-    ///         std::cout << "the optional contains the value: " << result << std::endl;
-    ///     })
-    /// @endcode
-    ///
-    template <typename Optional = ValueType, typename std::enable_if<is_optional<Optional>::value, int>::type = 0>
-    expected& and_then(const cxx::function_ref<void(typename Optional::type&)>& callable) noexcept;
+    template <typename ChainableType = ValueType,
+              typename std::enable_if<is_chainable<ChainableType>::value, int>::type = 0>
+    const expected& and_then(const cxx::function_ref<void(typename flatten<ChainableType>::type&)>& callable) const
```

Comment:
These methods provide different logic when the type is chainable, i.e., has an `and_then` method.